### PR TITLE
feat: conversation participants

### LINKED ISSUE #241: As an agent, I would like to know which all agents have participated in the conversation
- [x] Add participant table to store the conversation participants. Table references a conversation and, a contact, or a user. 
- [ ] Whenever a new user/contact replies to the conversation add them to the participant table.
- [x] Show the list of participants a new table in the sidebar
- [ ] tagging agents add thems to participants table
- [ ] ability to receive notifications from participating conversations

### PR DESCRIPTION
**Pull request by sojan-official**
Fixes #241
Fixes: https://github.com/chatwoot/product/issues/648

https://user-images.githubusercontent.com/73185/214910425-ca5b4cc1-1e4d-4c5b-b9c3-96db686d3ac9.mp4

### CODE REVIEW

**Comment by nithindavid** [comment]
@pranavrajs  @sojan-official I have updated the UI. https://www.loom.com/share/381665d68eb7404aa16f30dd92e058c2

File: .rubocop.yml [hunk 1]
```diff
@@ -12,13 +12,14 @@ Metrics/ClassLength:
...
     - 'app/models/contact.rb'
+    - 'app/models/user.rb'
     - 'app/mailers/conversation_reply_mailer.rb'
     - 'app/models/message.rb'
     - 'app/builders/messages/facebook/message_builder.rb'
     - 'app/controllers/api/v1/accounts/contacts_controller.rb'
     - 'app/controllers/api/v1/accounts/conversations_controller.rb'
     - 'app/listeners/action_cable_listener.rb'
-    - 'app/models/conversation.rb'
+    - 'app/models/user.rb'
```
Comment (pranavrajs) [inline .rubocop.yml:22]: Do we need this line?

File: app/javascript/dashboard/components/ChatList.vue [hunk 2]
```diff
@@ -352,6 +352,9 @@ export default {
       if (this.conversationType === 'mention') {
         return this.$t('CHAT_LIST.MENTION_HEADING');
       }
+      if (this.conversationType === 'participating') {
+        return 'Participating';
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:356]: Please use i18n.

File: app/finders/conversation_finder.rb [hunk 3]
```diff
@@ -76,9 +76,12 @@ def set_team
   end
 
   def find_all_conversations
-    if params[:conversation_type] == 'mention'
+    case params[:conversation_type]
+    when 'mention'
       conversation_ids = current_account.mentions.where(user: current_user).pluck(:conversation_id)
       @conversations = current_account.conversations.where(id: conversation_ids)
+    when 'participating'
+      @conversations = current_user.participating_conversations.where(account_id: current_account.id)
```
Comment (pranavrajs) [inline app/finders/conversation_finder.rb:84]: Do we need to push an update in the mobile app for this notification?

File: db/schema.rb [hunk 4]
```diff
@@ -561,6 +572,30 @@
...
+    t.string "last_name"
+    t.string "orders_count"
+    t.integer "contact_id"
+    t.integer "account_id"
+    t.string "customer_id"
+    t.integer "shopify_account_id"
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+  end
+
```
Comment (nithindavid) [inline db/schema.rb:598]:
```suggestion
```

File: db/schema.rb [hunk 5]
```diff
@@ -451,7 +462,7 @@
     t.text "attribute_description"
     t.jsonb "attribute_values", default: []
     t.index ["account_id"], name: "index_custom_attribute_definitions_on_account_id"
-    t.index ["attribute_key", "attribute_model", "account_id"], name: "attribute_key_model_index", unique: true
+    t.index ["attribute_key", "attribute_model"], name: "attribute_key_model_index", unique: true
```
Comment (nithindavid) [inline db/schema.rb:465]:
```suggestion
```

**Comment by sojan-official** [comment]
@pranavrajs  @nithindavid, if we concur on the requirements and the frontend changes, I will work out the pending backend changes and specs.

**Comment by pranavrajs** [comment]
-   **How does an agent get added to a conversation as a watcher?**
    -   Agent can add themselves to the watcher list manually.
    -   @mentions should add the agents mentioned to the watcher list.
    -   Assignment should automatically add the agent to the list.
-   **Who can remove the watchers?**
    -   An agent can remove themselves from the watcher list
    -   An admin can remove themselves or any other agent from the watcher list.
-   An agent should be part of the inbox to be added as a watcher.
-   When an agent is removed from an inbox, they should be removed from all the conversations they are watching on that inbox.
-   **A new notification type: Participating**
    -   All messages in the participating conversations should send a notification. (Incoming / Outgoing / Private / Activity)
    -   This can be split into multiple type if one config for every message type seems too much.

**Comment by pranavrajs** [comment]
@nithindavid Clicking on 1 other should show all the agents' avatars in the list. 

Screenshot 2022-11-09 at 7 32 11 PM

**Comment by pranavrajs** [comment]
**Updated from the discussion:** 

- Adding as a watcher permits the user to view the conversation even if there is no access to the inbox. This would enable the users to collaborate across the team. 

-   **How does an agent get added to a conversation as a watcher?**
    -   Agent can add themselves to the watcher list manually.
    -   @mentions should add the agents mentioned to the watcher list automatically.
    -   Assignment should automatically add the agent to the list.


-   **A new notification type: Participating**
    -   All messages in the participating conversations should send a notification. (Incoming / Outgoing / Private / Activity)
    -   This can be split into multiple type if one config for every message type seems too much.

**Comment by sojan-official** [comment]
Note:  enabling access to inboxes without permission is trickier than anticipated, as we have quite a few endpoints with permission checks and specs. Let's take it up as a follow-up. 

There is also the concern that some admins might not want agents to have permission to add another agent to a conversation to which the second agent doesn't have access.  

---- 

In this version, the add participants dropdown will only show other agents with access to this inbox.

**Comment by sojan-official** [comment]
Screenshot 2023-01-27 at 2 50 18 AM

@nithindavid, the sidebar looks a little off on this branch. 
everything else should be functional. Could you take a look?

**Comment by glauberssoutodossantos** [comment]
Hello guys,

Has this feature been abandoned? If yes, wouldn't it make more sense to discard/delete the PR? This PR is very interesting, but it's been almost 1 year since it's been here as open.

**Comment by sojan-official** [comment]
> Hello guys,
> 
> Has this feature been abandoned? If yes, wouldn't it make more sense to discard/delete the PR? This PR is very interesting, but it's been almost 1 year since it's been here as open.

It got delayed because of various priorities. But we are planning to get this merged by the upcoming release.

**Review by muhsin-k — COMMENTED** [review]
QA completed staging and working as expected.

File: app/listeners/notification_listener.rb [hunk 6]
```diff
@@ -28,50 +28,10 @@ def assignee_changed(event)
...
-      NotificationBuilder.new(
-        notification_type: 'conversation_mention',
-        user: User.find(user_id),
-        account: account,
-        primary_actor: message
-      ).perform
-    end
+    # Check if the message contains any mentions
+    Messages::MentionService.new(message: message).perform
+    Messages::NewMessageNotificationService.new(message: message).perform
```
[the remark this answers is not in the document]
Comment (sojan-official) [reply app/listeners/notification_listener.rb:35]: Screenshot 2023-01-30 at 3 42 46 PM

We have this notification specifically for new messages that we plan to depreciate down the lane. ( added this temporarily for participants as well )
Hence abstracted the logic to this service class which we want to get rid of when we get to that. ( moved the code to the class to satisfy rubocop ). We don't want it to be generic as it is to be removed in future.

**Comment by sojan-official** [comment]
@pranavrajs  waiting to see if there are any product comments before merging this.

**Comment by KarzounApps** [comment]
Game changer feature, Hope to officially see it in the next update.

**Pull request merged**